Enable users to update username on first comment

mario-galic opened the pull request:
## What does this change?

This allow users to set their username the first time they post a comment, that is, during discussion on-boarding process.

## What is the value of this and can you measure success?

Users that register via social sign in (Google, Facebook) have their `displayNames` automatically set to their real names, thus when they comment they are unaware that the comment will be posted under their real name. Currently there is no self-service mechanism to allow users to change their username. For this reason they often contact Userhelp to request account deletion. 

Also, `username/displayName` has been removed from all registration forms to improve conversion, since usernames are used only for commenting, so there is no need to have this extra field during Membership or Subscription registration flows:
* https://github.com/guardian/identity-frontend/pull/187
* https://github.com/guardian/identity-frontend/pull/195

## Does this affect other platforms - Amp, Apps, etc?

https://github.com/guardian/identity/pull/604

## Screenshots

![image](https://cloud.githubusercontent.com/assets/13835317/21546402/6ec1fb56-cdd6-11e6-9b4d-59d6119d2ca5.png)

Error path:

![image](https://cloud.githubusercontent.com/assets/13835317/21546432/9ef26252-cdd6-11e6-8d52-2c2fa5232c9f.png)

@markjamesbutler @thomaska @andrewfindlay



<!-- AB test? https://git.io/v1V0x -->
<!-- AMP question? https://git.io/v1V0p -->
<!-- Does this PR meet the contributing guidelines? https://git.io/v1VEJ -->

paperboyo commented:
This is great! Just a question: some usernames contain spaces in the current system (e.g. https://profile.theguardian.com/user/id/16400749). How’s that compatible with the screenshot where the spaces are said to not be allowed?

andrewfindlay commented:
@paperboyo I can't answer many questions on Github, I can answer this one.

Usernames cannot contain spaces, display names can. At the moment, anyone who signs up through social sign in gets given a display name based on their Facebook/Google name which doesn't have the same restrictions as a username. If they go on to comment, their display name is used. Once this change goes through, anyone with a display name will have to change it to a username when they first comment.

paperboyo commented:
So no space in comment names for newly signed up users. Clear, @andrewfindlay, thanks!

katebee commented on common/app/views/fragments/commentBox.scala.html:
```diff
@@ -24,6 +24,13 @@
+
+        <div class="d-comment-box__onboarding-username">
+            <p>Please choose your username under which you would like all your comments to show up:</p>
+            <input type="text" class="d-comment-box__onboarding-username-input" placeholder="Username: 6-20 characters, letters & numbers only, no spaces" value=""/>
```
The username requirements seem like key information; it may be better to add them as a label outside of the input area? 

Not all screen readers read placeholder text aloud:
https://www.nngroup.com/articles/form-design-placeholders/

mario-galic commented:
@katebee I've added a label and moved placeholder text out of the field to improve accessibility:

![image](https://cloud.githubusercontent.com/assets/13835317/21564317/1ed0fa12-ce84-11e6-852e-1e726ee724f4.png)

andrewfindlay commented:
@mario-galic I'd suggest two slight changes to the text, assuming there is space. Instead of:

Username: Must have 6-20 characters, letters & numbers only, no spaces.

I would suggest:

Username: Must be 6-20 characters, letters and/or numbers only, no spaces.

We do get the occasional user who thinks that they have to use numbers in their username when it says letters and numbers.

katebee commented on static/src/stylesheets/module/_discussion.scss:
```diff
@@ -1530,3 +1530,34 @@ $comment-recommend-button-size: 19px;
+}
+
+.d-comment-box__onboarding-username-label {
+    font-size: 14px;
```
maybe: `@include fs-textSans(3);` instead?

https://github.com/guardian/guss

katebee commented on static/src/stylesheets/module/_discussion.scss:
```diff
@@ -1530,3 +1530,34 @@ $comment-recommend-button-size: 19px;
+    margin: 0.2rem 0;
+    width: 70%;
+    margin-bottom: $gs-baseline;
+
```
Could values from the grid be used instead in padding and margin?

e.g. `padding: $gs-baseline/2 $gs-gutter;`

https://github.com/guardian/guss-grid-system

katebee commented:
(Someone with an eye for design please jump in and stop me...) Would it be beneficial to visually separate the user comment from the surrounding text? 

For example, adding background colour to `d-comment-box__onboarding-preview-body`:

<img width="663" alt="picture 432" src="https://cloud.githubusercontent.com/assets/8607683/21607172/3d3d2f68-d1ac-11e6-8e2c-eeb25a4a3c0f.png">

katebee commented:
If you want to make the error messages match, `identity/conf/messages.en` line 8 can be updated:

`error.usernameLength=Please enter a username between 6 and 20 characters`

katebee commented on static/src/javascripts/projects/common/modules/discussion/comment-box.js:
```diff
@@ -517,6 +551,29 @@ CommentBox.prototype.formatComment = function(formatStyle) {
+
+    var menuHeaderUsername = document.querySelector(".js-profile-info");
+    var discussionHeaderUsername = document.querySelector("._author_tywwu_16");
+    var firstCommentUsername = document.querySelector('[data-comment-author-id=' + '"' + userId + '"]');
```
- Single quotes on lines 568 and 569
- Qwery (`$`) could be used rather than `document.querySelector` (and it is already being included in the AMD):

```
var menuHeaderUsername = $('.js-profile-info')[0];
var discussionHeaderUsername = $('._author_tywwu_16')[0];
var firstCommentUsername = $('[data-comment-author-id=' + '"' + userId + '"]')[0];
```
...Additionally, my `discussionHeaderUsername` is updating, but not the `firstCommentUsername`This seems to be because `firstCommentUsername = Null`. I assume this is just an issue when running locally?

katebee commented:
<img width="716" alt="picture 436" src="https://cloud.githubusercontent.com/assets/8607683/21649052/3943c4a2-d298-11e6-928f-0e5a6e52cb7c.png">

crifmulholland commented:
👍  Nice work guys
